lib/src/Uncertainty/Algorithm/MetaModel/LinearModel/LinearModelAnalysis.cxx

```diff
@@ -260,21 +260,31 @@ Scalar LinearModelAnalysis::getFisherScore() const
   const UnsignedInteger size = residuals.getSize();
   // Get the number of parameter p
   const UnsignedInteger p = linearModelResult_.getCoefficients().getSize();
-  // Define RSS and SYY
```

Comment:
We should check if p>1 otherwise the divisions by (p-1) will produce strange results...

Reply:
@regislebrun Thanks for pointing this!

In fact it is much more complex that I thought

The correction `-1` is due to the intercept! In our API we can not handle it (e.g. we can not ensure that `intercept` is used in the provided basis)

In other libraries (`R` or `statmodels` for example), as we are handling `formula` we are aware of `intercept`/`no intercept`

An example in R:
```
library(faraway)
g1 <- lm(sr ~ pop15 - 1, data=savings)
g2 <- lm(sr ~ pop15 , data=savings)
g3 <- lm(sr ~ pop15 + pop75 -1 , data=savings)

summary(g1)
summary(g2)
summary(g3)
```

which returns :

```
Call:
lm(formula = sr ~ pop15 - 1, data = savings)

Residuals:
   Min     1Q Median     3Q    Max
-9.114 -3.828  2.010  5.313 14.498

Coefficients:
      Estimate Std. Error t value Pr(>|t|)
pop15  0.24445    0.02323   10.52  3.6e-14 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 5.952 on 49 degrees of freedom
Multiple R-squared:  0.6933,	Adjusted R-squared:  0.687
F-statistic: 110.7 on 1 and 49 DF,  p-value: 3.597e-14
```

```
Call:
lm(formula = sr ~ pop15 + pop75 - 1, data = savings)

Residuals:
    Min      1Q  Median      3Q     Max
-9.8004 -3.1565 -0.3508  2.5561 13.4738

Coefficients:
      Estimate Std. Error t value Pr(>|t|)
pop15  0.12884    0.02615   4.927 1.03e-05 ***
pop75  2.17074    0.36101   6.013 2.40e-07 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 4.542 on 48 degrees of freedom
Multiple R-squared:  0.825,	Adjusted R-squared:  0.8178
F-statistic: 113.2 on 2 and 48 DF,  p-value: < 2.2e-16
```

```
Call:
lm(formula = sr ~ pop15 + pop75, data = savings)

Residuals:
    Min      1Q  Median      3Q     Max
-8.7246 -2.7036 -0.1198  2.2825 10.3211

Coefficients:
            Estimate Std. Error t value Pr(>|t|)
(Intercept)  30.6277     7.4085   4.134 0.000146 ***
pop15        -0.4708     0.1468  -3.207 0.002414 **
pop75        -1.9341     1.0409  -1.858 0.069425 .
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.931 on 47 degrees of freedom
Multiple R-squared:  0.2617,	Adjusted R-squared:  0.2303
F-statistic: 8.332 on 2 and 47 DF,  p-value: 0.0007995
```

We notice that for both models, used `DoF` for first argument of the `Fisher` test is `basisSize - 1` if `intercept` is a component of the model.

Having a look at the data:
```
> savings$pop15
   [1] 29.35 23.32 23.80 41.89 42.19 31.72 39.74 44.75 46.64 47.64 24.42 46.31
  [13] 27.84 25.06 23.31 25.62 46.05 47.32 34.03 41.31 31.16 24.52 27.01 41.74
  [25] 21.80 32.54 25.95 24.71 32.61 45.04 43.56 41.18 44.19 46.26 28.96 31.94
  [37] 31.92 27.74 21.44 23.49 43.42 46.12 23.27 29.81 46.40 45.25 41.12 28.13
  [49] 43.69 47.20

> savings$pop75
 [1] 2.87 4.41 4.43 1.67 0.83 2.85 1.34 0.67 1.06 1.14 3.93 1.19 2.37 4.70 3.35
[16] 3.10 0.87 0.58 3.08 0.96 4.19 3.48 1.91 0.91 3.73 2.47 3.67 3.25 3.17 1.21
[31] 1.20 1.05 1.28 1.12 2.85 2.28 1.52 2.87 4.54 3.73 1.08 1.21 4.46 3.43 0.90
[46] 0.56 1.73 2.72 2.07 0.66

> savings$sr
 [1] 11.43 12.07 13.17  5.75 12.88  8.79  0.60 11.90  4.98 10.78 16.85  3.59
[13] 11.24 12.64 12.55 10.67  3.01  7.70  1.27  9.00 11.34 14.28 21.10  3.98
[25] 10.35 15.48 10.25 14.65 10.67  7.30  4.44  2.02 12.70 12.78 12.49 11.14
[37] 13.30 11.77  6.86 14.13  5.13  2.81  7.81  7.56  9.22 18.56  7.72  9.24
[49]  8.89  4.71
```

We implement the same models in `openturns`. We start with the full model:
```
import openturns as ot

pop15 = [29.35, 23.32, 23.80, 41.89, 42.19, 31.72, 39.74, 44.75, 46.64, 47.64, 24.42, 46.31,\
         27.84, 25.06, 23.31, 25.62, 46.05, 47.32, 34.03, 41.31, 31.16, 24.52, 27.01, 41.74,\
         21.80, 32.54, 25.95, 24.71, 32.61, 45.04, 43.56, 41.18, 44.19, 46.26, 28.96, 31.94,\
         31.92, 27.74, 21.44, 23.49, 43.42, 46.12, 23.27, 29.81, 46.40, 45.25, 41.12, 28.13,\
         43.69, 47.20]

pop75 = [2.87, 4.41, 4.43, 1.67, 0.83, 2.85, 1.34, 0.67, 1.06, 1.14, 3.93, 1.19, 2.37, 4.70, 3.35,\
         3.10, 0.87, 0.58, 3.08, 0.96, 4.19, 3.48, 1.91, 0.91, 3.73, 2.47, 3.67, 3.25, 3.17, 1.21,\
         1.20, 1.05, 1.28, 1.12, 2.85, 2.28, 1.52, 2.87, 4.54, 3.73, 1.08, 1.21, 4.46, 3.43, 0.90,\
         0.56, 1.73, 2.72, 2.07, 0.66]

sr = [11.43, 12.07, 13.17, 5.75, 12.88, 8.79,  0.60, 11.90,  4.98, 10.78, 16.85,  3.59,\
      11.24, 12.64, 12.55, 10.67,  3.01,  7.70,  1.27,  9.00, 11.34, 14.28, 21.10,  3.98,\
      10.35, 15.48, 10.25, 14.65, 10.67, 7.30 , 4.44 , 2.02, 12.70, 12.78, 12.49, 11.14,\
      13.30, 11.77,  6.86, 14.13,  5.13,  2.81,  7.81,  7.56,  9.22, 18.56,  7.72,  9.24,\
      8.89,  4.71]

data = ot.Sample(len(pop15), 0)
P = ot.Sample.BuildFromPoint(pop15)
data.stack(P)
P = ot.Sample.BuildFromPoint(pop75)
data.stack(P)
data.setDescription(["pop15", "pop75"])

Y = ot.Sample.BuildFromPoint(sr)
algo = ot.LinearModelAlgorithm(data, Y)
algo.run()
result = algo.getResult()
print(ot.LinearModelAnalysis(result))

# Model with pop15 + pop75 without intercept
# We do not run mode with pop15 only for the moment
f = ot.SymbolicFunction(["pop15", "pop75"], ["pop15"])
g = ot.SymbolicFunction(["pop15", "pop75"], ["pop75"])
basis = ot.Basis([f, g])
model = ot.LinearModelAlgorithm(data, basis, Y)
model.run()
print(ot.LinearModelAnalysis(model.getResult()))
```

The outputs are the following:
```
Basis( [[pop15,pop75]->[1],[pop15,pop75]->[pop15],[pop15,pop75]->[pop75]] )

Coefficients:
                       | Estimate    | Std Error   | t value     | Pr(>|t|)    |
--------------------------------------------------------------------------------
[pop15,pop75]->[1]     | 30.6277     | 7.40853     | 4.13411     | 0.000145698 |
[pop15,pop75]->[pop15] | -0.470843   | 0.146813    | -3.2071     | 0.0024142   |
[pop15,pop75]->[pop75] | -1.93413    | 1.04092     | -1.8581     | 0.0694254   |
--------------------------------------------------------------------------------

Residual standard error: 3.9307 on 47 degrees of freedom
F-statistic: 8.3318 ,  p-value: 0.00079955
---------------------------------
Multiple R-squared   | 0.261746 |
Adjusted R-squared   | 0.23033  |
---------------------------------

---------------------------------
Normality test       | p-value  |
---------------------------------
Anderson-Darling     | 0.974912 |
Cramer-Von Mises     | 0.980017 |
Chi-Squared          | 0.991468 |
Kolmogorov-Smirnov   | 0.999246 |
---------------------------------
```

```
Basis( [[pop15,pop75]->[pop15],[pop15,pop75]->[pop75]] )

Coefficients:
                       | Estimate    | Std Error   | t value     | Pr(>|t|)    |
--------------------------------------------------------------------------------
[pop15,pop75]->[pop15] | 0.128843    | 0.0261493   | 4.9272      | 1.03255e-05 |
[pop15,pop75]->[pop75] | 2.17074     | 0.361013    | 6.01292     | 2.39599e-07 |
--------------------------------------------------------------------------------

Residual standard error: 4.542 on 48 degrees of freedom
F-statistic: -0.31993 ,  p-value: 1
------------------------------------
Multiple R-squared   | -0.00670998 |
Adjusted R-squared   | -0.0276831  |
------------------------------------

---------------------------------
Normality test       | p-value  |
---------------------------------
Anderson-Darling     | 0.604944 |
Cramer-Von Mises     | 0.84934  |
Chi-Squared          | 0.739918 |
Kolmogorov-Smirnov   | 0.992737 |
---------------------------------
```
The model without intercept return a bad statistic ( < 0). We need to fix it.

Probably the "best" way to fix this is to rely on the `design` in `LinearModelResult` and activate an internal attribut `intercept` if one of the column is constant.
What do you think?
@regislebrun @jschueller @josephmure and others?

Reply:
yes, keep it simple

Reply:
do you want to do that in another PR after the RC or here ?

Reply:
here. I will do it this morning

Reply:
I agree with your idea to deal with the presence or absence of an intercept.

Reply:
Done. It was more complicated than expected. Reason is that many statistics required an update...